core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -304,15 +299,52 @@ class TiBatchWrite(@transient val df: DataFrame,
       }
     }
 
-    val deduplicatedTiRowRdd = deduplicateIfNecessary(tiRowRdd, tiTableInfo.isPkHandle)
-
     // get timestamp as start_ts
     val startTimeStamp = tiSession.getTimestamp
 
-    // TODO support insert ignore
-    val toBeCheckedRdd = getKeysNeedCheck(deduplicatedTiRowRdd, startTimeStamp)
-    checkConflictWithInsertKeys(toBeCheckedRdd)
-    val encodedTiRowRDD = generateRDDToBeInserted(toBeCheckedRdd)
+    // for partition table, we need calculate each row and tell which physical table
+    // that row is belong to.
+    // currently we only support upsert and insert.
+    val constraintCheckIsNeeded = handleCol != null || uniqueIndices.nonEmpty
+
+    val encodedTiRowRDD = if (constraintCheckIsNeeded) {
+      val wrappedRowRdd = if (tiTableInfo.isPkHandle) {
+        tiRowRdd.map { row =>
+          WrappedRow(row, extractHandleId(row))
+        }
+      } else {
+        val start = getAutoTableIdStart(tiRowRdd.count)
+        tiRowRdd.zipWithIndex.map { data =>
+          WrappedRow(data._1, data._2 + start)
+        }
+      }
+
+      val distinctWrappedRowRdd = deduplicate(wrappedRowRdd)
+
+      val rddDeletion = generateDataToBeRemovedRdd(distinctWrappedRowRdd, startTimeStamp)
```

Comment:
better to use one word, choose one between `Deletion` and `Remove`

Reply:
what if row duplicates in `rddDeletion`? should we deduplicate? pls add tests to ensure correctness.

Reply:
key in rdd deletion is unique.